src/Lean/Elab/BuiltinCommand.lean

```diff
@@ -223,6 +341,42 @@ private def replaceBinderAnnotation (binder : TSyntax ``Parser.Term.bracketedBin
   else
     return #[binder]
 
+/-- Declares a list of typed variables.
+
+Introduces variables that can be used in definitions within the same scope. When a definition
+mentions a variable, Lean will add it as an argument of the definition. This process is also able to
+infer typeclass parameters.
+
+Variable declarations have the same flexibility as regular function paramaters. In particular they
+can be explicit, implicit, or instance implicit (in which case they can be anonymous). This can be
+changed, for instance one can turn explicit variable `x` into an implicit one with `variable {x}`.
```

Comment:
IIRC it's not advisable to use this feature in the same `variable` command as declaring other variables; is it worth mentioning that here?

Reply:
@eric-wieser Do you have a quick description of this available?

Reply:
I don't have a description, but I have a pathological example:
```lean
class Foo (α : Type) where

def Bar (α : Type) [Foo α] := Nat

variable {α : Type} [Foo α]

-- now let's do a "binder update":

section
variable (α) (x : Bar α) -- fails; put your cursor before `x` to see why
end

section
variable (α)
variable (x : Bar α) -- same thing on two separate lines: ok
end
```

Reply:
So the short answer is "don't put them on the same line because you'll often get a confusing error message that leads you into even more of a mess".

Reply:
[No one knows](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/unexpected.20.60variable.60.20behaviour.3F/near/377995272) whether this behavior is a bug or by design.

Reply:
Update: I filed #2789, so I guess we just mention it in the docstring as justification for not mixing variables and binder updates.

Reply:
Update: I filed #2789, so I guess we just mention it in the docstring as justification for not mixing variables and binder updates.